oracle/src/lib.rs

```diff
@@ -1,30 +1,108 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use support::{decl_event, decl_module, decl_storage};
+mod default_combine_data;
+mod mock;
+mod operator_provider;
+mod tests;
+mod timestamped_value;
+
+pub use default_combine_data::DefaultCombineData;
+pub use operator_provider::OperatorProvider;
+use rstd::prelude::Vec;
+use rstd::result;
+use sr_primitives::traits::Member;
+use support::{decl_error, decl_event, decl_module, decl_storage, dispatch::Result, ensure, traits::Time, Parameter};
+use system::ensure_signed;
+pub use timestamped_value::TimestampedValue;
+pub use traits::{CombineData, OnNewData};
+
+type MomentOf<T> = <<T as Trait>::Time as Time>::Moment;
 
 pub trait Trait: system::Trait {
 	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+	type OnNewData: OnNewData<Self::Key, Self::Value>;
+	type OperatorProvider: OperatorProvider<Self::AccountId>;
+	type CombineData: CombineData<Self::Key, TimestampedValue<Self::Value, MomentOf<Self>>>;
+	type Time: Time;
+	type Key: Parameter + Member + Copy + Ord;
+	type Value: Parameter + Member + Copy + Ord;
 }
 
 decl_storage! {
 	trait Store for Module<T: Trait> as Oracle {
+		pub RawValues get(raw_values): double_map T::Key, blake2_256(T::AccountId) => Option<TimestampedValue<T::Value, MomentOf<T>>>;
+		pub HasUpdate get(has_update): map T::Key => bool;
+		pub Values get(values): map T::Key => Option<TimestampedValue<T::Value, MomentOf<T>>>;
+	}
+}
+
+decl_error! {
+	// Oracle module errors
+	pub enum Error {
+		NoPermission,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
 
+		pub fn feed_data(origin, key: T::Key, value: T::Value) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_feed_data(who, key, value).map_err(|e| e.into())
+		}
 	}
 }
 
 decl_event!(
 	pub enum Event<T> where
-		<T as system::Trait>::AccountId
+		<T as system::Trait>::AccountId,
+		<T as Trait>::Key,
+		<T as Trait>::Value,
 	{
-		Dummy(AccountId),
+		NewFeedData(AccountId, Key, Value),
 	}
 );
 
-decl_module! {
-	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
-		fn deposit_event() = default;
+impl<T: Trait> Module<T> {
+	pub fn read_raw_values(key: &T::Key) -> Vec<TimestampedValue<T::Value, MomentOf<T>>> {
```

Comment:
Since `T::Key` has `Copy` bound, `&` is not needed, same to other functions where `T::Key` is a parameter, except `combine_data`.